simulation/replication/replication_simulation_test.go

```diff
@@ -0,0 +1,323 @@
+func TestReplicationSimulation(t *testing.T) {
+	flag.Parse()
+
+	logf(t, "Starting Replication Simulation")
+
+	logf(t, "Sleeping for 30 seconds to allow services to start/warmup")
+	time.Sleep(30 * time.Second)
+
+	// load config
+	simCfg := mustLoadReplSimConf(t)
+
+	// initialize cadence clients
+	for clusterName := range simCfg.Clusters {
+		simCfg.mustInitClientsFor(t, clusterName)
+	}
+
+	mustRegisterDomain(t, simCfg)
+
+	// wait for domain data to be replicated.
+	// TODO: make domain replication interval configurable and reduce it
+	time.Sleep(10 * time.Second)
+
+	// initialize workers
+	for clusterName := range simCfg.Clusters {
+		simCfg.mustInitWorkerFor(t, clusterName)
+	}
+
+	startTime := time.Now().UTC()
+	logf(t, "Simulation start time: %v", startTime)
+	g := &errgroup.Group{}
+	var maxAt time.Duration
+	for _, op := range simCfg.Operations {
+		op := op
+		if op.At > maxAt {
+			maxAt = op.At
+		}
+		switch op.Type {
+		case ReplicationSimulationOperationStartWorkflow:
+			g.Go(func() error {
+				return startWorkflow(t, op, simCfg, startTime)
+			})
+		case ReplicationSimulationOperationFailover:
+			g.Go(func() error {
+				return failover(t, op, simCfg, startTime)
+			})
+		case ReplicationSimulationOperationValidate:
+			g.Go(func() error {
+				return validate(t, op, simCfg, startTime)
+			})
+		default:
+			require.Failf(t, "unknown operation type", "operation type: %s", op.Type)
+		}
+	}
+
+	logf(t, "Waiting for all operations to complete.. Latest operation time is %v", startTime.Add(maxAt))
+	g.Wait()
+
+	// Print the test summary.
+	// Don't change the start/end line format as it is used by scripts to parse the summary info
+	executionTime := time.Since(startTime)
+	testSummary := []string{}
+	testSummary = append(testSummary, "Simulation Summary:")
+	testSummary = append(testSummary, fmt.Sprintf("Simulation Duration: %v", executionTime))
+	testSummary = append(testSummary, "End of Simulation Summary")
+	fmt.Println(strings.Join(testSummary, "\n"))
+}
+
+func startWorkflow(
+	t *testing.T,
+	op *Operation,
+	simCfg *ReplicationSimulationConfig,
+	startTime time.Time,
+) error {
+	t.Helper()
+
+	waitForOpTime(t, op, startTime)
+
+	logf(t, "Starting workflow: %s", op.WorkflowID)
+
+	// TODO: start workflow in specified cluster
+
+	return nil
+}
+
+func failover(
+	t *testing.T,
+	op *Operation,
+	simCfg *ReplicationSimulationConfig,
+	startTime time.Time,
+) error {
+	t.Helper()
+
+	waitForOpTime(t, op, startTime)
+
+	logf(t, "Failing over to cluster: %s", op.NewActiveCluster)
+
+	// TODO: perform failover
+
+	return nil
+}
+
+func validate(
+	t *testing.T,
+	op *Operation,
+	simCfg *ReplicationSimulationConfig,
+	startTime time.Time,
+) error {
+	t.Helper()
+
+	waitForOpTime(t, op, startTime)
+
+	logf(t, "Performing valiaton")
+
+	// TODO: perform validation
+
+	return nil
+}
+
+func waitForOpTime(t *testing.T, op *Operation, startTime time.Time) {
+	t.Helper()
+	<-time.After(startTime.Add(op.At).Sub(time.Now().UTC()))
```

Comment:
curious as to the rationale of putting it in UTC, though i suppose in the buildkite logs it's easier to follow

Reply:
In this case timezone of time object passed to `Sub()` doesn't matter. The underlying subtraction will do proper subtraction considering the timezone difference. So I could have omitted it.